src/connection/mod.rs

```diff
@@ -73,101 +88,87 @@ pub fn delete(c: &Connection) -> Result<(), String> {
-                                      NM_SERVICE_INTERFACE,
-                                      "ActiveConnections")
-        .inner::<&Vec<dbus::MessageItem>>()
-        .unwrap()
-        .iter()
-        .map(|p| dbus_path_to_string(p.inner::<&dbus::Path>().unwrap().to_owned()))
-        .collect::<Vec<_>>();
-    // How can we avoid this ^ collect - it gets turned back into an iterator below.
-
-    // Get a vector containing settings paths for all active connections
-    let settings_paths = active_paths.iter()
-        .map(|p| {
-            dbus_path_to_string(dbus_property!(NM_SERVICE_MANAGER,
-                                               p,
-                                               NM_ACTIVE_INTERFACE,
-                                               "Connection")
-                .inner::<&dbus::Path>()
-                .unwrap()
-                .to_owned())
-        });
-
-    // Pre-set the state as it won't get changed unless the passed in connection.path is present
-    let mut state = ConnectionState::Deactivated;
-
-    // Loop over the active paths and settings paths
-    // If the passed in connection.path is equal to the settings path the state is
-    // retrieved using the active connection path
-    for (active_path, settings_path) in active_paths.iter().zip(settings_paths) {
-        if settings_path == connection.path {
-            state = ConnectionState::from(dbus_property!(NM_SERVICE_MANAGER,
-                                                         active_path,
-                                                         NM_ACTIVE_INTERFACE,
-                                                         "State")
-                .inner::<u32>()
-                .unwrap());
-            break;
-        }
-    }
+#[test]
+fn test_enable_disable_functions() {
+    let connections = list().unwrap();
+
+    // Note - replace "TP-LINK_2.4GHz_9BDD8F" with one of your configured connections to test
+    let mut connection =
+        connections.iter().filter(|c| c.ssid == "TP-LINK_2.4GHz_9BDD8F").nth(0).unwrap().clone();
+
+    assert!(connection.state == ConnectionState::Activated ||
+            connection.state == ConnectionState::Deactivated);
+
+    if connection.state == ConnectionState::Activated {
+        disable(&mut connection, 10).unwrap();
+        assert_eq!(ConnectionState::Deactivated, connection.state);
 
-    Ok(state)
+        enable(&mut connection, 10).unwrap();
+        assert_eq!(ConnectionState::Activated, connection.state);
+    } else {
+        enable(&mut connection, 10).unwrap();
+        assert_eq!(ConnectionState::Activated, connection.state);
+
+        disable(&mut connection, 10).unwrap();
+        assert_eq!(ConnectionState::Deactivated, connection.state);
+    }
 }
 
-fn get(path: dbus::Path) -> Result<Connection, String> {
+fn get_connection(path: dbus::Path) -> Result<Connection, String> {
     let mut connection = Connection { path: dbus_path_to_string(path), ..Default::default() };
```

Comment:
Unfortunately its not, trait implementations must be inside the same crate as the type `dbus::Path`